# Host-virus cophylogenetic analysis

This repository contains data and code used during the analyses resulting in the article entitled *"Intrahost speciations and host switches shaped the evolution of herpesviruses
"*.


## Codes

The directories `1_host_timetree`, `2_jane_nexus` and `3_jane_cli` contain the following bash and python codes used in our study:

* `timetree.py`
* `add_traits.py`
* `nexus2jane.py`
* `parse_jane_files.py`
* `runJane.sh`
* `parse_jane_files.py`


## Data

The directories `sequences` and `trees` contain raw data and phylogenies used in our study.

* `XXXX_degap.fasta`      → Degapped multiple sequence alignment of herpesvirus proteins (UL27 and UL30).
* `XXXX_full-length.fasta` → Unaligned sequences of herpesvirus proteins  (UL27 and UL30).
* `host_treetime.tree`   → Host tree built using data from [timetree.org](https://timetree.org/).
* `virus_beast_mcc.tree`  → Herpesvirus tree inferred using BEAST v1.10.4.


## Author

* **Anderson Brito** - [WebPage](https://andersonbrito.github.io/) - andersonfbrito@gmail.com

## License

This project is licensed under the MIT License.

<!---
--->
